common/types/activation.go

```diff
@@ -188,16 +174,30 @@ func NewActivationTx(
 	return atx
 }
 
-// SetMetadata sets ATXMetadata.
-func (atx *ActivationTx) SetMetadata() {
-	atx.Target = atx.TargetEpoch()
-	atx.MsgHash = BytesToHash(atx.HashInnerBytes())
+// ATXSigMsg is the data of ActivationTx that is signed.
+type ATXSigMsg struct {
+	ID           ATXID
+	PublishEpoch EpochID
```

Comment:
the epoch ID needs to be the first field for protocol update purpose

Reply:
we need to write down somewhere how we are going to support update for every type. i thought that version for malf proofs is decoded from the outermost layer id

we made malf proof to be `layer_id, malf_proof_enum`. if we need upgrades, we decode layer id, determine version of the enum and then decode the rest

Reply:
@dshulyak is only for the metadata or also for the ATX itself?

I will update the `struct` to look like this:

```go
type ATXMetadata struct {
	Layer LayerID
	Hash  Hash32
}
```

The epoch can be derived from the `LayerID` when checking for malfeasance, and this way it is generic enough to be used for all 3 types of proof we have at the moment.

Reply:
@countvonzero for protocol update purposes can we use to the `LayerID` in the `MalfeasanceProof` type? https://github.com/spacemeshos/go-spacemesh/blob/remove-metadata-from-atx/common/types/malfeasance.go#L22

This way we can do upgrades without having to limit ourselves to a specific layout of the `XXXMetadata` structs.

Reply:
i think that atx metadata should continue using epoch and order doesn't matter. as a rule we should use only the first field of the object (in this case MalfeasanceProof) for the version, everything else after that field is decoded based on the version 

```
layer := decodeLayer(buf)
switch layerToVersion(layer) {
     case 0:
     case 1:       
}
```